### PR TITLE
Test/seanimam azure monitor

### PR DESCRIPTION
**Pull request by seanimam**
## Description

> Adds new files to the telemetry-generator package that enable performance benchmarking to be sent to Azure App Insights.

## Breaking Changes

> None

## Reviewer Guidance

>

### CODE REVIEW

File: tools/telemetry-generator/package.json [hunk 1]
```diff
@@ -28,9 +28,11 @@
     "tsc": "tsc"
   },
   "dependencies": {
+    "@azure/identity": "^3.1.3",
```
Comment (alexvy86) [inline tools/telemetry-generator/package.json:31]: Seems like we don't need this one?

File: tools/pipelines/test-perf-benchmarks.yml [hunk 2]
```diff
@@ -197,6 +198,19 @@ stages:
...
 
+        - task: Bash@3
+          displayName: Send Execution Time Performance Benchmark Measurments to Azure App Insights
+          inputs:
+            targetType: 'inline'
+            workingDirectory: $(toolAbsolutePath)
+            script: |
+              echo "Writing performance benchmark output to Azure App Insights..."
+              ls -laR ${{ variables.consolidatedTestsOutputFolder }};
+              node --require @ff-internal/aria-logger bin/run appInsights --handlerModule $(toolAbsolutePath)/dist/handlers/appInsightsExecutionTimeTestHandler.js --dir '${{ variables.consolidatedTestsOutputFolder }}' --connectionString '$(fluid-app-insights-connection-string)';
```
Comment (alexvy86) [inline tools/pipelines/test-perf-benchmarks.yml:209]:
```suggestion
              node bin/run appInsights --handlerModule $(toolAbsolutePath)/dist/handlers/appInsightsExecutionTimeTestHandler.js --dir '${{ variables.consolidatedTestsOutputFolder }}' --connectionString '$(fluid-app-insights-connection-string)';
```
The new steps don't need to write anything to Aria/Kusto so we don't need that `--require`. Let's remove it from the other new steps too.

File: tools/pipelines/test-perf-benchmarks.yml [hunk 3]
```diff
@@ -485,6 +512,32 @@ stages:
               ls -la ${{ variables.memoryUsageTestOutputFolder }};
               node --require @ff-internal/aria-logger bin/run --handlerModule $(toolAbsolutePath)/dist/handlers/memoryUsageTestHandler.js --dir ${{ variables.memoryUsageTestOutputFolder }};
 
+        - task: Bash@3
+          displayName: Send Excecution Time Perf Benchmark Measurments to Azure App Insights
```
Comment (alexvy86) [inline tools/pipelines/test-perf-benchmarks.yml:516]:
```suggestion
          displayName: Send Execution Time Perf Benchmark Measurments to Azure App Insights
```

File: tools/pipelines/test-perf-benchmarks.yml [hunk 4]
```diff
@@ -485,6 +512,32 @@ stages:
...
+          inputs:
+            targetType: 'inline'
+            workingDirectory: $(toolAbsolutePath)
+            script: |
+              echo "Writing memory usage performance benchmark output to Azure App Insights..."
```
Comment (alexvy86) [inline tools/pipelines/test-perf-benchmarks.yml:521]: This echo and the one in line 534 should be swapped

File: tools/telemetry-generator/src/handlers/appInsightsExecutionTimeTestHandler.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * This handler uses the TemeletryClient as its logger which is from the 'applicationinsights' Azure package.
+ */
+module.exports = async function handler(fileData, logger) {
```
Comment (alexvy86) [inline tools/telemetry-generator/src/handlers/appInsightsExecutionTimeTestHandler.ts:9]: In both handlers, let's rename `logger` to `telemetryClient` to make this clearer, and in the comment we can say that these handlers are supposed to work in conjunction with the appInsights command, which should provide the correct object in that second parameter.

File: tools/telemetry-generator/src/commands/appInsights.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        }),
+    };
+
+    static examples = [
+        {
+            command: "$ node bin/run --handlerModule /path/to/my/module.js --dir /path/to/my/files",
+            description: "Process files from /path/to/my/files and all its subfolders, using the handler at " +
+                "/path/to/my/module.js",
+        },
+    ];
```
Comment (alexvy86) [inline tools/telemetry-generator/src/commands/appInsights.ts:40]: Update this with the extra pieces for the appInsights command.

File: tools/telemetry-generator/src/commands/appInsights.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            handler = (await import(flags.handlerModule)).default;
+        } catch (err) {
+            exitWithError(`Unexpected error importing specified handler module.\n${err}`);
+        }
+
+        if (typeof handler !== "function") {
+            exitWithError("Handler module does not have a function as its default export");
+        }
+
+        console.log(`obtained connectionstring: ${flags.connectionString}`)
```
Comment (alexvy86) [inline tools/telemetry-generator/src/commands/appInsights.ts:59]: I'd say probably remove? Locally we're passing it to the command, and in ADO it gets replaced with *** for security.

File: tools/telemetry-generator/src/commands/appInsights.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        appInsight.setup(flags.connectionString)
+        .start();
+        const logger = appInsight.defaultClient;
```
Comment (alexvy86) [inline tools/telemetry-generator/src/commands/appInsights.ts:63]: We can rename this to telemetryClient as well, to keep in line with the terminology used by the AppInsights SDK.

**Review by alexvy86 — COMMENTED** [review]
Mostly minor things, but looks good!

**Comment by alexvy86** [comment]
@seanimam can you also check how many different entries we currently have in Kusto? This clean number in AppInsights feels very suspicious:
image

File: tools/telemetry-generator/src/commands/appInsights.ts (hunk 8 above)
Comment (seanimam) [reply tools/telemetry-generator/src/commands/appInsights.ts]: Done in recent commit # 84045

File: tools/telemetry-generator/src/commands/appInsights.ts (hunk 7 above)
Comment (seanimam) [reply tools/telemetry-generator/src/commands/appInsights.ts]: Yeah this was for debugging. Removed in recent commit # 84045

File: tools/telemetry-generator/src/commands/appInsights.ts (hunk 6 above)
Comment (seanimam) [reply tools/telemetry-generator/src/commands/appInsights.ts]: Done in recent commit # 84045

File: tools/telemetry-generator/src/handlers/appInsightsExecutionTimeTestHandler.ts (hunk 5 above)
Comment (seanimam) [reply tools/telemetry-generator/src/handlers/appInsightsExecutionTimeTestHandler.ts]: Done in recent commit # 84045

File: tools/pipelines/test-perf-benchmarks.yml (hunk 3 above)
Comment (seanimam) [reply tools/pipelines/test-perf-benchmarks.yml]: Fixed in recent commit # 84045

File: tools/pipelines/test-perf-benchmarks.yml (hunk 4 above)
Comment (seanimam) [reply tools/pipelines/test-perf-benchmarks.yml]: Fixed in recent commit # 84045

File: tools/pipelines/test-perf-benchmarks.yml (hunk 2 above)
Comment (seanimam) [reply tools/pipelines/test-perf-benchmarks.yml]: Done in recent commit # 84045

File: tools/telemetry-generator/package.json (hunk 1 above)
Comment (seanimam) [reply tools/telemetry-generator/package.json]: Yeah this was added when I was trying to emit metrics directly to azure monitor and I was going to use the identity package to obtain an oAuth token.

**Review by alexvy86 — APPROVED** [review]
Code looks good. @seanimam do you have any updates on the metric count comparison with Kusto?

**Comment by seanimam** [comment]
> Code looks good. @seanimam do you have any updates on the metric count comparison with Kusto?

I have manually confirmed all metrics that the tool is attempting to send is successfully making it to app insights without the need for any rate limiting

**Comment by seanimam** [comment]
> @seanimam can you also check how many different entries we currently have in Kusto? This clean number in AppInsights feels very suspicious: image

Discussed this offline and after adding a flush command and manually checking we are confirming all metrics making it to app insights

**Pull request merged**